### ISSUE TITLE
[Bug] MasterDetailPage doesn't work as tab item TabBar

### ISSUE THREAD

**Issue by evoynov** [xamarin/Xamarin.Forms#7346]
### Description

I need to make the flow below:
App - 3 tabs - the second tab is MasterDetailPage with (Dynamic) ListView on MasterPage where every listview item opens a different page on detail side.

Using Xamarin.Forms it was reached by TabbedPage + MasterDetailPage.
Using Xamarin Shell I used TabBar element and the same MasterDetailPage, but it's represented by black area on iOS (but processes in the background still work, checked in debug) and crashes on Android with stacktrace:

"Specified cast is not valid."
  at Xamarin.Forms.Platform.Android.VisualElementExtensions.IsAttachedToRoot (Xamarin.Forms.VisualElement Element) [0x00047] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\VisualElementExtensions.cs:44 \n  at Xamarin.Forms.Platform.Android.AppCompat.MasterDetailContainer.AddChildView (Xamarin.Forms.VisualElement childView) [0x00066] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\MasterDetailContainer.cs:93 \n  at Xamarin.Forms.Platform.Android.MasterDetailContainer.set_ChildView (Xamarin.Forms.VisualElement value) [0x0002e] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\Renderers\\MasterDetailContainer.cs:47 \n  at Xamarin.Forms.Platform.Android.AppCompat.MasterDetailPageRenderer.<UpdateDetail>g__Update|79_0 () [0x00022] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\MasterDetailPageRenderer.cs:410 \n  at Xamarin.Forms.Platform.Android.AppCompat.MasterDetailPageRenderer.UpdateDetail () [0x0000d] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\MasterDetailPageRenderer.cs:399 \n  at Xamarin.Forms.Platform.Android.AppCompat.MasterDetailPageRenderer.Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00135] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\MasterDetailPageRenderer.cs:158 \n  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Support.V4.App.FragmentManager fragmentManager, Android.Content.Context context) [0x00031] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\Platform.cs:393 \n  at Xamarin.Forms.Platform.Android.AppCompat.FragmentContainer.OnCreateView (Android.Views.LayoutInflater inflater, Android.Views.ViewGroup container, Android.OS.Bundle savedInstanceState) [0x00008] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\FragmentContainer.cs:56 \n  at Xamarin.Forms.Platform.Android.AppCompat.ShellFragmentContainer.OnCreateView (Android.Views.LayoutInflater inflater, Android.Views.ViewGroup container, Android.OS.Bundle savedInstanceState) [0x00011] in D:\\a\\1\\s\\Xamarin.Forms.Platform.Android\\AppCompat\\ShellFragmentContainer.cs:38 \n  at Android.Support.V4.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_inflater, System.IntPtr native_container, System.IntPtr native_savedInstanceState) [0x00020] in <21e4bb539f4a49b58a498749b367d2a8>:0 \n  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.39(intptr,intptr,intptr,intptr,intptr)

### Steps to Reproduce

1. Create Xamarin Shell project
2. Organize first level of navigation as TabBar
3. Make one of the tabs as MasterDetailPage descendant
4. Open this tab on the device

### Expected Behavior
![photo5431358580042345400](https://user-images.githubusercontent.com/3380639/64080737-66ce4900-cd00-11e9-95d4-16def66dc1ca.jpg)

### Actual Behavior
![photo5431594786063756156](https://user-images.githubusercontent.com/3380639/64080740-6b92fd00-cd00-11e9-99ab-e896641d03c3.jpg)

### Basic Information

- Version with issue: 4.2.0.709249
- Last known good version: 3.6.0.344457
- IDE: Visual Studio Community 2019 for Mac Version 8.2.5 (build 5)
- Platform Target Frameworks: 
  - iOS:  8.0
  - Android: 8.1

**Comment by samhouts** [comment]
@evoynov  Can you please attach a small project that demonstrates this issue? Thanks!

**Comment by evoynov** [comment]
@samhouts published a minimal version of my project as a reproduction: https://github.com/evoynov/MasterDetailShellIssue

**Comment by samhouts** [comment]
@evoynov Thank you for the sample. I requested one because you indicated that this was previously working on 3.6.0, which was surprising! What you have described is not a supported scenario. It does not look like it was previously working, so I have removed the regression status.

MasterPage is not supported as a Tab in Shell. Only ContentPages may be used.

With Shell, you are essentially getting a MasterPage with a TabbedPage as the Detail, which is the opposite of what you're looking for. The Master page (which is called Flyout in Shell) is meant to be global for the whole application, but since you're using a TabBar and you have the Flyout disabled, you can "customize" it for your needs. 

In the `OnAppearing` for the ContentPage you create for Tab2, enable the Flyout (`Shell.Current.FlyoutBehavior = FlyoutBehavior.Flyout`). And in your definition of Tab2 in the AppShell, you can add your custom MenuItems that represent your Table of Contents. Then you would attach commands to those menu items that would navigate or change the page content.

[7346.zip](https://github.com/xamarin/Xamarin.Forms/files/3592291/7346.zip)

I'm going to keep this issue open, however, because the error message should be more informative when you try to use something other than a content page.

Thanks!

**Comment by evoynov** [comment]
@samhouts thank you a lot for a deep explanation, but I've tried your version of my application and It doesn't work. Do I understand correctly and the marked area (look at the image below) should contain "burger" menu that would open flyout with menu items? Tried on iPhone XR 12.4 simulator and on X 13.1 beta device.

![photo5456548430089399082](https://user-images.githubusercontent.com/3380639/64570867-9fa69780-d36a-11e9-90b5-5e3fb25fd31b.jpg)

**Comment by samhouts** [comment]
@evoynov For the moment, you'll need to have the flyout always enabled. We've just fixed a bug that prevents you from enabling the flyout on a page by page basis. That should be resolved in 4.3.0. #7460 and #7459. Thanks for your patience!

**Comment by timmi4sa** [comment]
@samhouts Thanks for helping resolve this. Will the fixes #7460 and #7459 be appearing in the nightly builds shortly? We are at crossroads: switch back to X.Forms 3, or stick with the broken Shell.

**Comment by samhouts** [comment]
@timmi4sa Yes! Both were merged, and they should be in the next nightly build. They will also be in the next 4.3.0 prerelease, which should be ready in the next week. Thanks!
